### PR TITLE
[DualStack] Support NSG and clean LBs

### PR DESCRIPTION
**Pull request by lzhecheng**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
* Support NSG and clean LBs for dualstack
* Support related UTs for dualstack
* Refactor
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
* Support NSG and clean LBs for dualstack
* Support related UTs for dualstack
* Refactor
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: tests/e2e/network/node.go [hunk 1]
```diff
@@ -197,7 +197,7 @@ var _ = Describe("Azure node resources", Label(utils.TestSuiteLabelNode), func()
 
 		utils.Logf("getting all NICs of VMSS VMs")
 		for _, vmssVM := range vmssVMs {
-			utils.Logf("Checking %d VMSS VM %q", vmssVM.Name)
+			utils.Logf("Checking %d VMSS VM %q", *vmssVM.Name)
```
Comment (jwtty) [inline tests/e2e/network/node.go:200]: 2 args?

File: pkg/provider/azure_vmssflex.go [hunk 2]
```diff
@@ -989,7 +989,9 @@ func (fs *FlexScaleSet) ensureBackendPoolDeletedFromNode(vmssFlexVMNameMap map[s
 			nics[nicName] = nic
 		}
 	}
+	nicUpdaterLock := "nicUpdaterLock"
```
Comment (jwtty) [inline pkg/provider/azure_vmssflex.go:992]: +1 atomic.Store() may be better than lock?

File: pkg/provider/azure_privatelinkservice.go [hunk 3]
```diff
@@ -44,24 +44,29 @@ func (az *Cloud) reconcilePrivateLinkService(
...
+	createPLS := wantPLS && serviceRequiresPLS(service)
+	isDualStack := isServiceDualStack(service)
 	if isIPv6 {
-		klog.V(2).Infof("IPv6 is not supported for private link service, skip reconcilePrivateLinkService for service(%s)", service)
-		return nil
+		if isDualStack || !createPLS {
+			klog.V(2).Infof("IPv6 is not supported for private link service, skip reconcilePrivateLinkService for service(%s)", serviceName)
+			return nil
+		}
+		if !isDualStack {
```
Comment (jwtty) [inline pkg/provider/azure_privatelinkservice.go:65]: We don't need this if at all?

File: tests/e2e/network/node.go (hunk 1 above)
Comment (lzhecheng) [reply tests/e2e/network/node.go]: Updated.

File: pkg/provider/azure_privatelinkservice.go (hunk 3 above)
Comment (lzhecheng) [reply pkg/provider/azure_privatelinkservice.go]: Updated.

File: pkg/provider/azure_vmssflex.go (hunk 2 above)
Comment (lzhecheng) [reply pkg/provider/azure_vmssflex.go:992]: Will update in the other PR and rebase.

File: pkg/provider/azure_loadbalancer_backendpool.go [hunk 4]
```diff
@@ -70,11 +70,11 @@ func (bc *backendPoolTypeNodeIPConfig) EnsureHostsInPool(service *v1.Service, no
 }
 
 func isLBBackendPoolsExisting(lbBackendPoolNames map[bool]string, bpName *string) (found, isIPv6 bool) {
-	if strings.EqualFold(pointer.StringDeref(bpName, ""), lbBackendPoolNames[false]) {
+	if strings.EqualFold(pointer.StringDeref(bpName, ""), lbBackendPoolNames[IsIPv4]) {
```
Comment (MartinForReal) [inline pkg/provider/azure_loadbalancer_backendpool.go:73]: I'm wondering if following code snips will be simpler. 
```go
if strings.EqualFold(pointer.StringDeref(bpName, ""){
 if lbBackendPoolNames[IsIPv4]){
}
}
```

File: pkg/provider/azure_loadbalancer_backendpool.go [hunk 5]
```diff
@@ -135,10 +135,10 @@ func (bc *backendPoolTypeNodeIPConfig) CleanupVMSetFromBackendPoolByCondition(sl
 			})
 		}
 		if v4Enabled {
-			findBackendpoolToBeDeleted(false)
+			findBackendpoolToBeDeleted(IsIPv4)
```
Comment (MartinForReal) [inline pkg/provider/azure_loadbalancer_backendpool.go:138]: can we call findBackendpoolToBeDeleted directly?

File: pkg/provider/azure_standard.go [hunk 6]
```diff
@@ -296,8 +296,8 @@ func getBackendPoolName(clusterName string, isIPv6 bool) string {
 // getBackendPoolNames returns the IPv4 and IPv6 backend pool names.
 func getBackendPoolNames(clusterName string) map[bool]string {
 	return map[bool]string{
-		false: getBackendPoolName(clusterName, false),
-		true:  getBackendPoolName(clusterName, true),
+		IsIPv4: getBackendPoolName(clusterName, IsIPv4),
```
Comment (MartinForReal) [inline pkg/provider/azure_standard.go:299]: can we define a enum type for IsIPv4 and IsIPv6? I'm not sure these two variables are mutually exclusive.

Comment (lzhecheng) [reply pkg/provider/azure_standard.go:299]: IsIPv4 is actually NotIPv6, so they are mutually exclusive.

File: tests/e2e/network/node.go [hunk 7]
```diff
@@ -197,7 +197,7 @@ var _ = Describe("Azure node resources", Label(utils.TestSuiteLabelNode), func()
 
 		utils.Logf("getting all NICs of VMSS VMs")
 		for _, vmssVM := range vmssVMs {
-			utils.Logf("Checking %d VMSS VM %q", vmssVM.Name)
+			utils.Logf("Checking VMSS VM %q", *vmssVM.Name)
```
Comment (MartinForReal) [inline tests/e2e/network/node.go:200]: shall we safely refer to this variable securely by invoking to. String?

Comment (lzhecheng) [reply tests/e2e/network/node.go]: Updated.

File: pkg/provider/azure_loadbalancer.go [hunk 8]
```diff
@@ -68,10 +68,10 @@ func (az *Cloud) existsPip(clusterName string, service *v1.Service) bool {
 		return existingPip
 	}
 
-	if v4Enabled && !existsPipSingleStack(false) {
+	if v4Enabled && !existsPipSingleStack(IsIPv4) {
```
Comment (nilo19) [inline pkg/provider/azure_loadbalancer.go:71]: I think a better name would be IPVersionIPv4. IsIPv4 sounds like a variable, not a constant.

File: pkg/provider/azure_loadbalancer.go [hunk 9]
```diff
@@ -693,19 +714,25 @@ func (az *Cloud) selectLoadBalancer(clusterName string, service *v1.Service, exi
 	return selectedLB, existsLb, nil
 }
 
-func (az *Cloud) getServiceLoadBalancerStatus(service *v1.Service, lb *network.LoadBalancer) (status *v1.LoadBalancerStatus, fipConfig *network.FrontendIPConfiguration, err error) {
+// getServiceLoadBalancerStatus returns LB status for the Service.
+// Before DualStack support, old logic takes the first ingress IP as non-additional one
+// and the second one as additional one. With DualStack support, the second IP may be
+// the IP of another IP family so the new logic returns two variables.
+func (az *Cloud) getServiceLoadBalancerStatus(service *v1.Service, lb *network.LoadBalancer) (status *v1.LoadBalancerStatus, lbIPsNoAdditionalPIPs []string, fipConfigs []*network.FrontendIPConfiguration, err error) {
```
Comment (nilo19) [inline pkg/provider/azure_loadbalancer.go:721]: is noAdditional means primary?

File: pkg/provider/azure_loadbalancer.go (hunk 8 above)
Comment (lzhecheng) [reply pkg/provider/azure_loadbalancer.go]: Updated.

File: pkg/provider/azure_utils.go [hunk 10]
```diff
@@ -34,6 +34,11 @@ import (
 	"sigs.k8s.io/cloud-provider-azure/pkg/consts"
 )
 
+const (
+	IsIPv6 bool = true
```
[the remark this answers is not in the document]
Comment (lzhecheng) [reply pkg/provider/azure_utils.go]: Updated.

File: pkg/provider/azure_loadbalancer.go (hunk 9 above)
Comment (lzhecheng) [reply pkg/provider/azure_loadbalancer.go:721]: It means compared to ingress IPs in `status`, `lbIPsNoAdditionalPIPs` doesn't include those IPs from `service.beta.kubernetes.io/azure-additional-public-ips` annotation.

Comment (nilo19) [reply pkg/provider/azure_loadbalancer.go:721]: Will it be better to replace to primary?

Comment (lzhecheng) [reply pkg/provider/azure_loadbalancer.go]: Yes, `NoAdditional` seems too straightforward. Replaced.

**Pull request merged**